### ISSUE TITLE
Create an irregular polygon shaped cylinder and use it to clip a mesh

### ISSUE THREAD

**Issue by craigmillernz** [pyvista/pyvista-support#50]
I'd like to be able to clip a mesh by an irregular shaped polygon cylinder. This polygon could be defined by the perimeter of a set of datapoints (ie like observations in a geophysical survey), or it could be defined manually using coordinates of the top and bottom of the cylinder.
I would then use this polygon cylinder surface as input into the `dataset.clip_surface()` filter.  The intention is to clip a geophysical model (which is a rectangular cube) to the actual data  measurement point distribution, which is a random shape.

**Comment by banesullivan** [comment]
So just to clarify and follow up on some points from Slack - the `clip_surface` filter requires a closed ("watertight") mesh to work - we'd need to create a `PolyData` mesh of the defined surface that encloses all the data in the mesh that you'd like to extract.

Here is an example of how we might create a cyclider and clip a gridded dataset to extract a non-uniform ROI.

FYI - I'm experiencing a significant issue/bug with the `clip_surface` filter when using custom, irregular surfaces that will need to be mitigated, so I don't have an immediate solution for you when it comes to clipping with custom surfaces - I'll report back

## Example A

Used PyVista' cylinder helper method to create a cylinder `PolyData` object that is closed and clip a `UniformGrid` with that surface:

```py
import pyvista as pv
from pyvista import examples

# The gridded dataset to clip
mesh = examples.load_uniform()

# The surface that bounds the ROI
surf = pv.Cylinder(center=mesh.center, direction=(1,2,1),
                   height=15, radius=3.3)

p = pv.Plotter(notebook=0) 
p.add_mesh(surf, color='red', opacity=0.5,)
p.add_mesh(mesh, show_edges=True)
p.show()
```

![2019-08-14 17 28 24](https://user-images.githubusercontent.com/22067021/63063275-39d50600-beb9-11e9-93eb-22fa435b8475.gif)


```py
# Perform the clip
clipped = mesh.clip_surface(surf, compute_distance=True)

p = pv.Plotter(notebook=0) 
p.add_mesh(mesh.outline(), color=True)
p.add_mesh(clipped, show_edges=True)
p.show()
```

![2019-08-14 17 28 43](https://user-images.githubusercontent.com/22067021/63063300-4d806c80-beb9-11e9-8052-ca7d6d92a7ec.gif)

**Comment by banesullivan** [comment]
Also, the `clip_box` filter now allows a `PolyData` box to be passed which should address this.

ref https://github.com/pyvista/pyvista/pull/433 which landed on `v0.23.0`
